### PR TITLE
Java: CWE-918: Server side request forgery

### PR DESCRIPTION
**Pull request by random-debug**
This query detects instances where untrusted user input may be used to instantiate a server side connection to a node. 

If network request are made using `java.net.URL` constructed directly from a user-provided value, an attacker might be able to bypass security restrictions such as external firewalls and execute request on internal services by providing URLs like `http://127.0.0.1:8081/admin`

This PR only considers network requests made using `URL.openConnection` method. There is already a query to detect network request made using `URL.openStream` in `java/ql/src/experimental/Security/CWE/CWE-036`. 

This PR also includes test cases, their expected output, and qhelp file for the query.

### CODE REVIEW

**Comment by intrigus-lgtm** [comment]
Besides #3452 there was also a similar query:
https://github.com/github/codeql/pull/2899 that is partly similar.
In the case of #2899 it looked not for SSRF but local file disclosure due to user-controlled url parameters.

**Comment by random-debug** [comment]
Hello @intrigus-lgtm,

There are certain defects specific to SSRF that can be detected using CodeQL queries. Some examples are :-

1. Redirect to blacklisted URL host attacks
    
    ``` java
    List<String> BLACKLIST = Arrays.asList(
            "localhost",
            "redis.internal.com"
    );

    URL url = new URL(request.getParameter("target"));
    if (BLACKLIST.contains(url.getHost().toLowerCase())) {
        throw new Exception("URL host is not allowed");
    }

    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
    // connection.setInstanceFollowRedirects(false); <-- redirect should be disabled
    connection.connect();
    ```
    
    By default `HttpURLConnection` will follow redirects. Redirects should be disabled by calling `connection.setInstanceFollowRedirects(false)`. If it is not disabled as shown in above code snippet, an attacker can bypass this blacklist by initially providing URL of a attacker controlled server which is not in blacklist and the server will send redirect response with `Location` header set to a URL host that is in blacklist.

2. DNS rebinding attacks due to Time of check Time of use defect. This type of defect is very common in java.
    
    ``` java
    List<String> BLACKLIST = Arrays.asList(
            "127.0.0.1",
            "192.168.1.1"
    );

    URL url = new URL(request.getParameter("target"));

    // <-- Time of check DNS resolution
    InetAddress address = InetAddress.getByName(url.getHost());
    if (BLACKLIST.contains(address.getHostAddress())) {
        throw new Exception("URL host is not allowed");
    }

    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
    connection.setInstanceFollowRedirects(false);

    // <-- Time of use DNS resolution
    connection.connect();
    ```

Since these defects are specific to SSRF, I think a dedicated query would be better than modifying #2899. The above mentioned defects are not detected in the current PR, but I was hoping to address them after a basic query for SSRF is merged. 

Also PR #3454 addresses lot more use cases than this one. But to reduce false positives, I have included some optimisations in the query like
1. It is OK to append user controlled input as suffix to `spec` parameter of a URL constructor. For example `new URL("http://domain.com/webhook?target=" + request.getParameter("target"))` is considered OK.
2. It is OK to add user controlled input as prefix to `host` parameter of a URL constructor. For example `new URL("https", request.getParameter("subdomain") + ".safe-domain.com", "webhook")` is considered OK.

I can send a new PR with the above optimisations (if they are OK) after #3454 is merged

**Comment by intrigus-lgtm** [comment]
@random-debug 
Just to be clear, I'm not affiliated with lgtm.com or github or Microsoft.

**Comment by pwntester** [comment]
Thanks for the submission @random-debug 

> Since these defects are specific to SSRF, I think a dedicated query would be better than modifying #2899. The above mentioned defects are not detected in the current PR, but I was hoping to address them in a new PR after a basic query for SSRF is merged.

That would make an interesting query

> I can send a new PR with the above optimisations (if they are OK) after #3454 is merged

That would be appreciated

For completeness and increasing the scope of the query please consider adding support for popular HTTP clients such as:
- Java 11 Http client
- Apache Commons HttpClient
- Apache Commons HttpComponents
- Google HTTP Client Library 
- JAX-RS client

**Comment by porcupineyhairs** [comment]
Just fyi,my PR (#3454) does support the Apache and java 11 sinks. I haven't added Google http client or jax-rs libraries though. 

So @random-debug, you can add those after #3454 is merged.

File: java/ql/src/experimental/Security/CWE/CWE-918/RequestForgery.ql [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @name Uncontrolled data used in network request
+ * @description Sending network requests with user-controlled data allows for request forgery attacks.
+ * @kind path-problem
+ * @problem.severity error
+ * @id java/request-forgery
```
Comment (aibaars) [inline java/ql/src/experimental/Security/CWE/CWE-918/RequestForgery.ql:6]: If I am not mistaken this query only addresses server-side request forgery and not others such as cross-site request forgery. Therefore I think we should make the `@id` more precise. I'd suggest `java/server-side-request-forgery` .

**Comment by random-debug** [comment]
Hi @pwntester @porcupineyhairs,
As discussed, I have improved the scope of this query by supporting Google HTTP client, JAX-RS and Spring REST template libraries.

**Comment by random-debug** [comment]
Closing this PR since most of the use cases detected by this query are handled by PRs #3454 and #3720

**Pull request closed without merging**